🌊 Streams: Add streams feature privilege

flash1293 opened the pull request:
This PR adds a new feature for streams to control whether the UI and API is available.

Changes:
* Add the feature with two privileges (`show` and `manage_assets`)
* Can be configured with the classic `none`/`read`/`all` so it's automatically aligned with serverless editor/viewer permissions
  * None also means the app is not shown - to do this, the existing `status$` observable also looks for at least the `streams.show` capability
* Only guards changes to the linked dashboards - changes to the Elasticsearch level are still delegated to the Elasticsearch-level permissions of the user
   * This happens on the UI level (disabled button and dashboard selection on the dashboard page)
   * and on the API level (all endpoints that can change linked dashboards require the permission)


# Questions

* Not sure about the name `manage_assets` - maybe it should be something else
* Not sure about how the queries stuff should work - @kdelemme anything we should do in this area?

<!--ONMERGE {"backportTargets":["8.19"]} ONMERGE-->

flash1293 committed:
add streams feature privilege

flash1293 committed:
fix

elasticmachine commented:
Pinging @elastic/obs-ux-logs-team (Team:obs-ux-logs)

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/styled_components_mapping'

klacabane commented on x-pack/platform/plugins/shared/streams_app/public/components/stream_detail_dashboards_view/index.tsx:
```diff
@@ -182,6 +190,7 @@ export function StreamDetailDashboardsView({
             <EuiButton
               data-test-subj="streamsAppStreamDetailAddDashboardButton"
               iconType="plusInCircle"
+              disabled={!canLinkAssets}
```
@patpscal should a button be disabled or hidden when user does not have necessary permissions ?

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/capture_oas_snapshot   --include-path /api/status   --include-path /api/alerting/rule/   --include-path /api/alerting/rules   --include-path /api/actions   --include-path /api/security/role   --include-path /api/spaces   --include-path /api/streams   --include-path /api/fleet   --include-path /api/dashboards   --include-path /api/saved_objects/_import   --include-path /api/saved_objects/_export   --include-path /api/alerting/maintenance_window --update'

kdelemme commented:
I had something similar for the sig event rule as well: https://github.com/elastic/kibana/pull/218173/files#diff-e6de66f96ca0b0f93249d07dc3c1ead5a5d79700a1ffc5e59ebe09a5d5c7b205R83-R131
I'll be able to add the alerting feature on top of your setup

jloleysens reviewed (approved):
Addition of feature's plugin LGTM

kdelemme commented on x-pack/platform/plugins/shared/streams/server/plugin.ts:
```diff
@@ -72,6 +76,37 @@ export class StreamsPlugin
+            all: [],
+            read: [],
+          },
+          api: [],
+          ui: [STREAMS_FEATURE_PRIVILEGES.show],
```
Does it mean we never enforce privileges on read APIs, and instead delegate to es? And for "write" api, we do check for the privilege?

klacabane replied:
Wondering if we should have base `read` privilege on all apis not already gated ? There's something off in not being able to read the UI but still being able to call the read apis

flash1293 replied:
> And for "write" api, we do check for the privilege?

Only for the ones that change the dashboard links.

> Wondering if we should have base read privilege on all apis not already gated ? There's something off in not being able to read from the UI but still being able to call the read apis

It's not adding anything they couldn't do directly talking to Elasticsearch, but I think you are right - it's weird that these other APIs are still available. I'm going to add that.

kibanamachine committed:
[CI] Auto-commit changed files from 'make api-docs'

klacabane reviewed (approved):
LGTM - tried with All, Read and None privileges

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/295418)
* Commit: e8ba84f042489745aa25a55330e57df9a1212bb7

### Failed CI Steps
* [FTR Configs #25](https://buildkite.com/elastic/kibana-pull-request/builds/295418#01966330-179e-4e26-bc31-4a7e613ce435)
* [FTR Configs #58](https://buildkite.com/elastic/kibana-pull-request/builds/295418#01966330-17b9-4143-b9cc-b60f25d15c85)
* [FTR Configs #74](https://buildkite.com/elastic/kibana-pull-request/builds/295418#01966330-17c6-4bc5-8241-8cfed34c8448)
* [FTR Configs #80](https://buildkite.com/elastic/kibana-pull-request/builds/295418#01966330-17cb-4925-bfa2-02c89d008650)
* [FTR Configs #96](https://buildkite.com/elastic/kibana-pull-request/builds/295418#01966330-17d8-45a7-9c0f-f80d780df802)
* [FTR Configs #127](https://buildkite.com/elastic/kibana-pull-request/builds/295418#01966330-1802-4f2c-92d7-f25a40b7a830)
* [FTR Configs #127](https://buildkite.com/elastic/kibana-pull-request/builds/295418#01966356-83b3-4407-a69a-08099c173463)
* [FTR Configs #128](https://buildkite.com/elastic/kibana-pull-request/builds/295418#01966330-1803-43d4-9efd-4c1d8d3dbafe)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/295418#01966330-179e-4e26-bc31-4a7e613ce435) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/295418/jobs/01966330-179e-4e26-bc31-4a7e613ce435/artifacts/01966358-ecea-4eb0-a182-2f606861f3b3) FTR Configs #<span></span>25 / features Features /api/features with trial license should return a full feature set
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/295418#01966330-1803-43d4-9efd-4c1d8d3dbafe) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/295418/jobs/01966330-1803-43d4-9efd-4c1d8d3dbafe/artifacts/01966358-f407-45eb-a9ef-34aca8dc6913) FTR Configs #<span></span>128 / security (basic license) Privileges GET /api/security/privileges should return a privilege map with all known privileges, without actions
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/295418#01966330-17c6-4bc5-8241-8cfed34c8448) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/295418/jobs/01966330-17c6-4bc5-8241-8cfed34c8448/artifacts/01966357-2cd9-4192-8d5e-b8a785785a45) FTR Configs #<span></span>74 / security Privileges GET /api/security/privileges should return a privilege map with all known privileges, without actions
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/295418#01966330-17cb-4925-bfa2-02c89d008650) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/295418/jobs/01966330-17cb-4925-bfa2-02c89d008650/artifacts/01966358-a148-4c18-b89f-adbd5a725ab8) FTR Configs #<span></span>80 / spaces api with security create superuser from the default space when solution is specified should return 200
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/295418#01966330-17d8-45a7-9c0f-f80d780df802) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/295418/jobs/01966330-17d8-45a7-9c0f-f80d780df802/artifacts/01966358-62e0-44b7-8f1f-5628752898b4) FTR Configs #<span></span>96 / spaces api with security create superuser from the default space when solution is specified should return 200
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/295418#01966330-17b9-4143-b9cc-b60f25d15c85) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/295418/jobs/01966330-17b9-4143-b9cc-b60f25d15c85/artifacts/01966355-d294-45f7-853f-1c685f5f1303) FTR Configs #<span></span>58 / spaces api without security create from the default space when solution is specified should return 200
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/295418#01966330-1802-4f2c-92d7-f25a40b7a830) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/295418/jobs/01966330-1802-4f2c-92d7-f25a40b7a830/artifacts/01966356-6f86-45a8-9847-bc227d4b75c5) FTR Configs #<span></span>127 / spaces telemetry data Verify disabledFeatures telemetry payloads includes only disabledFeatures findings
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/295418#01966356-83b3-4407-a69a-08099c173463) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/295418/jobs/01966356-83b3-4407-a69a-08099c173463/artifacts/0196635c-14b4-4e3d-b3c7-b3fc04d2a2df) FTR Configs #<span></span>127 / spaces telemetry data Verify disabledFeatures telemetry payloads includes only disabledFeatures findings

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/295418","number":295418,"commit":"e8ba84f042489745aa25a55330e57df9a1212bb7"}],"number":295418}
buildkite-pr-comment-->

flash1293 committed:
Merge branch 'main' into flash1293/streams-permissions

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/295617)
* Commit: 4d79e68355bb346bea8c3f18d570f517acabdaf7

### Failed CI Steps
* [FTR Configs #16](https://buildkite.com/elastic/kibana-pull-request/builds/295617#019666b7-877b-43e3-ab24-7779e79cd5b7)
* [FTR Configs #20](https://buildkite.com/elastic/kibana-pull-request/builds/295617#019666b7-877e-4fb3-a3a2-2f5632577613)
* [FTR Configs #30](https://buildkite.com/elastic/kibana-pull-request/builds/295617#019666b7-8786-491a-8ef6-699e270815de)
* [FTR Configs #56](https://buildkite.com/elastic/kibana-pull-request/builds/295617#019666b7-879b-49bb-a4a5-27e818c195c3)
* [FTR Configs #101](https://buildkite.com/elastic/kibana-pull-request/builds/295617#019666b7-87c2-4ad2-98e3-82ba228bc590)
* [FTR Configs #129](https://buildkite.com/elastic/kibana-pull-request/builds/295617#019666b7-87da-434a-96a5-1df62c1fb1ce)
* [FTR Configs #129](https://buildkite.com/elastic/kibana-pull-request/builds/295617#019666de-a16b-4822-92b8-077b87424dca)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/295617#019666b7-877e-4fb3-a3a2-2f5632577613) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/295617/jobs/019666b7-877e-4fb3-a3a2-2f5632577613/artifacts/019666de-f2ce-4370-8d4a-b787fde5c0ba) FTR Configs #<span></span>20 / features Features /api/features with trial license should return a full feature set
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/295617#019666df-0754-47a5-9a71-83e1a3eab4c4) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/295617/jobs/019666df-0754-47a5-9a71-83e1a3eab4c4/artifacts/019666e4-f214-4cfa-b78f-7c2034e5c9f7) FTR Configs #<span></span>20 / features Features /api/features with trial license should return a full feature set
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/295617#019666b7-87c2-4ad2-98e3-82ba228bc590) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/295617/jobs/019666b7-87c2-4ad2-98e3-82ba228bc590/artifacts/019666df-8387-445c-b4ab-5d4bc0cdc28d) FTR Configs #<span></span>101 / security Privileges GET /api/security/privileges should return a privilege map with all known privileges, without actions
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/295617#019666b7-877b-43e3-ab24-7779e79cd5b7) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/295617/jobs/019666b7-877b-43e3-ab24-7779e79cd5b7/artifacts/019666dc-26bf-4686-a5b4-f6647bd888de) FTR Configs #<span></span>16 / spaces api with security create superuser from the default space when solution is specified should return 200
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/295617#019666b7-879b-49bb-a4a5-27e818c195c3) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/295617/jobs/019666b7-879b-49bb-a4a5-27e818c195c3/artifacts/019666de-7437-4adc-9178-58ca12814244) FTR Configs #<span></span>56 / spaces api with security create superuser from the default space when solution is specified should return 200
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/295617#019666b7-8786-491a-8ef6-699e270815de) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/295617/jobs/019666b7-8786-491a-8ef6-699e270815de/artifacts/019666de-b748-42c6-bd57-f24d43904fc5) FTR Configs #<span></span>30 / spaces api without security create from the default space when solution is specified should return 200
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/295617#019666b7-87da-434a-96a5-1df62c1fb1ce) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/295617/jobs/019666b7-87da-434a-96a5-1df62c1fb1ce/artifacts/019666de-5916-4f03-9c29-a573dc1cbe9a) FTR Configs #<span></span>129 / spaces telemetry data Verify disabledFeatures telemetry payloads includes only disabledFeatures findings
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/295617#019666de-a16b-4822-92b8-077b87424dca) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/295617/jobs/019666de-a16b-4822-92b8-077b87424dca/artifacts/019666e4-d224-4277-abf7-9c6ee2e1ed23) FTR Configs #<span></span>129 / spaces telemetry data Verify disabledFeatures telemetry payloads includes only disabledFeatures findings

### History
* :broken_heart: [Build #295418](https://buildkite.com/elastic/kibana-pull-request/builds/295418) failed e8ba84f042489745aa25a55330e57df9a1212bb7

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/295617","number":295617,"commit":"4d79e68355bb346bea8c3f18d570f517acabdaf7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/295418","number":295418,"commit":"e8ba84f042489745aa25a55330e57df9a1212bb7"}],"number":295617}
buildkite-pr-comment-->

flash1293 committed:
Merge remote-tracking branch 'upstream/main' into flash1293/streams-permissions

flash1293 committed:
fix tests

flash1293 committed:
rename

flash1293 committed:
Merge branch 'flash1293/streams-permissions' of github.com:flash1293/kibana into flash1293/streams-permissions

flash1293 commented:
@klacabane @kdelemme Could you take another look? I added `read` and `write` permissions for the API and guarded all the API endpoints with it - this is a second layer on top of the Elasticsearch-level permissions checks we still need to do.

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/capture_oas_snapshot   --include-path /api/status   --include-path /api/alerting/rule/   --include-path /api/alerting/rules   --include-path /api/actions   --include-path /api/security/role   --include-path /api/spaces   --include-path /api/streams   --include-path /api/fleet   --include-path /api/dashboards   --include-path /api/saved_objects/_import   --include-path /api/saved_objects/_export   --include-path /api/alerting/maintenance_window --update'

kibanamachine committed:
[CI] Auto-commit changed files from 'make api-docs'

flash1293 committed:
remove exclusive suite

kdelemme commented on x-pack/platform/plugins/shared/streams/common/constants.ts:
```diff
@@ -6,3 +6,10 @@
 
 export const ASSET_VERSION = 1;
+
+export const STREAMS_FEATURE_PRIVILEGES = {
+  assets: 'manage_assets',
+  read: 'read',
+  write: 'manage_streams',
+  show: 'show',
+} as const;
```
Should we differentiate UI vs API privileges? 
I think API privileges require to be unique across all plugins' API privileges, so probably need to use read_stream, manage_streams and manage_assets.

Do we need the distinction between manage_streams and manage_assets for the APIs?

I would think this is enough:
```
all: { api: ["read_stream", "manage_stream"], ui: ["show", "manage"] }
read: { api: ["read_stream"], ui: ["show"] }
```

jeramysoucy replied:
I see these are used for both API privileges and UI capabilities. API privileges should have unique names globally (there is a guide here https://docs.elastic.dev/kibana-dev-docs/key-concepts/security-api-authorization#naming-conventions-for-privileges).

It probably makes sense to split these out, API and UI. Also, just a nit - keeping the value and name the same or similar.
```suggestion
export const STREAMS_FEATURE_API_PRIVILEGES = {
  assets: 'manage_stream_assets',
  read: 'read_stream',
  manage: 'manage_stream', // or write: 'write_stream',
} as const;

export const STREAMS_FEATURE_UI_CAPABILITIES = {
  manage: 'manage',
  show: 'show',
} as const;
```

kdelemme commented on x-pack/platform/plugins/shared/streams/server/routes/content/route.ts:
```diff
@@ -50,9 +51,7 @@ const exportContentRoute = createServerRoute({
-      enabled: false,
-      reason:
-        'This API delegates security to the currently logged in user and their Elasticsearch permissions.',
+      requiredPrivileges: [STREAMS_FEATURE_PRIVILEGES.assets, STREAMS_FEATURE_PRIVILEGES.read],
```
This means assets _and_ read are both required, which can only happen for the "all" access privilege, therefore just checking for assets is enough

jeramysoucy replied:
Alternatively, if you want to be able to access this route with either privilege you can use
`requiredPrivileges: [{ anyRequired: ['<privilege_1>', '<privilege_2>'] }],`

flash1293 replied:
simplified to just read

kdelemme commented on x-pack/platform/plugins/shared/streams/server/routes/internal/streams/processing/route.ts:
```diff
@@ -76,9 +75,7 @@ export const processingSuggestionRoute = createServerRoute({
-      enabled: false,
-      reason:
-        'This API delegates security to the currently logged in user and their Elasticsearch permissions.',
+      requiredPrivileges: [STREAMS_FEATURE_PRIVILEGES.read],
```
is read the right privilege for these processing APIs?

flash1293 replied:
This endpoint won't write anything, it's just hitting the LLM for suggestions. I think it can be useful even if the user doesn't have write permissions for streams to test around with different configurations

kdelemme commented on x-pack/platform/plugins/shared/streams/server/routes/queries/route.ts:
```diff
@@ -80,12 +79,9 @@ const upsertQueryRoute = createServerRoute({
-      enabled: false,
-      reason:
-        'This API delegates security to the currently logged in user and their Elasticsearch permissions.',
+      requiredPrivileges: [STREAMS_FEATURE_PRIVILEGES.read],
```
should it be manage_streams?

flash1293 replied:
yes, done

kdelemme commented on x-pack/platform/plugins/shared/streams/server/routes/streams/enablement/route.ts:
```diff
@@ -60,9 +59,7 @@ export const disableStreamsRoute = createServerRoute({
-      enabled: false,
-      reason:
-        'This API delegates security to the currently logged in user and their Elasticsearch permissions.',
+      requiredPrivileges: [STREAMS_FEATURE_PRIVILEGES.assets, STREAMS_FEATURE_PRIVILEGES.write],
```
We can use one or the other, since they are both only available to the "all" privilege access

jeramysoucy replied:
Pattern for this would be: `requiredPrivileges: [{ anyRequired: ['<privilege_1>', '<privilege_2>'] }],`

kdelemme replied:
I agree, but I wonder if we need to have these two privileges since they are only given when the feature access control is set to "All". The user either has these two privileges, or none. But this also goes back to my other thought: do we need both assets and write privileges or just a "manage_stream" privilege is enough?

flash1293 replied:
I left it in because I thought it's better to be explicit so we can give more fine-grained permissions later on, but that's actually not a good reason to do things. I will keep it simple and just go for read and manage

flash1293 replied:
Removed `assets`

kdelemme reviewed (commented):
I think the API privileges need to be unique (and read might not be). Also some APIs are requiring read privilege when they are updating resources (queries, processing, etc..) is it expected?

I think we can simplify the API privileges to just a read_stream and manage_stream, and the ui privileges to show and manage. what do you think?

jeramysoucy commented on x-pack/platform/plugins/shared/streams/server/routes/internal/streams/management/route.ts:
```diff
@@ -29,9 +30,7 @@ export const sampleStreamRoute = createServerRoute({
-      enabled: false,
-      reason:
-        'This API delegates security to the currently logged in user and their Elasticsearch permissions.',
+      requiredPrivileges: [STREAMS_FEATURE_PRIVILEGES.read],
```
Is read the correct privilege here?

flash1293 replied:
It's not, fixed

jeramysoucy reviewed (changes requested):
Thanks for adding authz to the streams routes! 🚀 
I left a few suggestions, and also some questions regarding the access level for some of the routes - I don't know the details of how these routes should work, but just wanted to double check that the correct privilege is being specified.

flash1293 committed:
Merge remote-tracking branch 'upstream/main' into flash1293/streams-permissions

flash1293 committed:
fixes

flash1293 commented:
Thanks for the reviews @klacabane @kdelemme and @jeramysoucy 

I adjusted as follows:
* Get rid of the assets privilege and consolidate into a more coarse `manage` - as long as we don't have the need we might as well keep it simple
* Layer the Kibana level privilege and the Elasticsearch level one and require the user to have both to allow performing actions - the Elasticsearch-level checks were already in place, this is merging it with the Kibana ones
* Fixed the wrong privilege levels for the content pack and query routes (please double check whether everything makes sense for you now in your areas)

Having this in place, I wonder whether we even need sub-features for content packs and queries - any strong opinons @LucaWintergerst ? Seems like a good one to only add if it actually becomes a need.

flash1293 commented on x-pack/platform/plugins/shared/streams/server/plugin.ts:
```diff
@@ -72,6 +76,41 @@ export class StreamsPlugin
+            STREAMS_FEATURE_PRIVILEGES.read,
+            STREAMS_FEATURE_PRIVILEGES.write,
+          ],
+          ui: [STREAMS_FEATURE_PRIVILEGES.show, STREAMS_FEATURE_PRIVILEGES.assets],
```
Done

flash1293 commented on x-pack/platform/plugins/shared/streams/server/routes/content/route.ts:
```diff
@@ -134,9 +133,7 @@ const importContentRoute = createServerRoute({
-      enabled: false,
-      reason:
-        'This API delegates security to the currently logged in user and their Elasticsearch permissions.',
+      requiredPrivileges: [STREAMS_FEATURE_PRIVILEGES.read],
```
It's not, updated

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/capture_oas_snapshot   --include-path /api/status   --include-path /api/alerting/rule/   --include-path /api/alerting/rules   --include-path /api/actions   --include-path /api/security/role   --include-path /api/spaces   --include-path /api/streams   --include-path /api/fleet   --include-path /api/dashboards   --include-path /api/saved_objects/_import   --include-path /api/saved_objects/_export   --include-path /api/alerting/maintenance_window --update'

kibanamachine committed:
[CI] Auto-commit changed files from 'make api-docs'

kdelemme reviewed (approved):
LGTM, thanks for the changes. One note on a custom hook, but can be done later when if needed

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/296155)
* Commit: e5b052a0c49221305bfec9dac31de81e38bccf2a

### Failed CI Steps
* [FTR Configs #18](https://buildkite.com/elastic/kibana-pull-request/builds/296155#01966cd4-1726-4af9-8291-722e942788d8)
* [FTR Configs #46](https://buildkite.com/elastic/kibana-pull-request/builds/296155#01966cd4-1736-4d02-91fa-3f1138c659a6)
* [FTR Configs #106](https://buildkite.com/elastic/kibana-pull-request/builds/296155#01966cd4-175c-487e-b7cc-548d811af3c4)
* [FTR Configs #128](https://buildkite.com/elastic/kibana-pull-request/builds/296155#01966cd4-176b-41d1-851f-71fa40ae424f)
* [FTR Configs #128](https://buildkite.com/elastic/kibana-pull-request/builds/296155#01966cfc-3816-457e-8e4b-dda028557725)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/296155#01966cd4-176b-41d1-851f-71fa40ae424f) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/296155/jobs/01966cd4-176b-41d1-851f-71fa40ae424f/artifacts/01966cfc-241a-49a2-85eb-fca786cb9fe2) FTR Configs #<span></span>128 / security (basic license) Privileges GET /api/security/privileges should include sub-feature privileges when respectlicenseLevel is false
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/296155#01966cfc-3816-457e-8e4b-dda028557725) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/296155/jobs/01966cfc-3816-457e-8e4b-dda028557725/artifacts/01966d02-afe9-48ce-8df8-3d5debd1b5e6) FTR Configs #<span></span>128 / security (basic license) Privileges GET /api/security/privileges should include sub-feature privileges when respectlicenseLevel is false
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/296155#01966cd4-1726-4af9-8291-722e942788d8) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/296155/jobs/01966cd4-1726-4af9-8291-722e942788d8/artifacts/01966cf9-352d-49eb-a095-f264eb152a3f) FTR Configs #<span></span>18 / spaces api with security get all superuser can access all spaces from default undefined purpose should return 200 when referencing the default space implicitly
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/296155#01966cd4-1736-4d02-91fa-3f1138c659a6) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/296155/jobs/01966cd4-1736-4d02-91fa-3f1138c659a6/artifacts/01966cfe-223b-4ee7-8051-abd4d980b7f3) FTR Configs #<span></span>46 / spaces api with security get all superuser can access all spaces from default undefined purpose should return 200 when referencing the default space implicitly
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/296155#01966cd4-175c-487e-b7cc-548d811af3c4) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/296155/jobs/01966cd4-175c-487e-b7cc-548d811af3c4/artifacts/01966cff-d5f7-4263-9a9f-23f987ac5bd7) FTR Configs #<span></span>106 / spaces api without security get all can access all spaces from default undefined purpose should return 200 when referencing the default space implicitly

### History
* :broken_heart: [Build #295890](https://buildkite.com/elastic/kibana-pull-request/builds/295890) failed 726d4976625c3b4ab963859a56a69614e2a19383
* :broken_heart: [Build #295861](https://buildkite.com/elastic/kibana-pull-request/builds/295861) failed 0491ca46ce4e1c54aaaeeb8831cb904e0a786b71
* :broken_heart: [Build #295617](https://buildkite.com/elastic/kibana-pull-request/builds/295617) failed 4d79e68355bb346bea8c3f18d570f517acabdaf7
* :broken_heart: [Build #295418](https://buildkite.com/elastic/kibana-pull-request/builds/295418) failed e8ba84f042489745aa25a55330e57df9a1212bb7

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/296155","number":296155,"commit":"e5b052a0c49221305bfec9dac31de81e38bccf2a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/295890","number":295890,"commit":"726d4976625c3b4ab963859a56a69614e2a19383"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/295861","number":295861,"commit":"0491ca46ce4e1c54aaaeeb8831cb904e0a786b71"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/295617","number":295617,"commit":"4d79e68355bb346bea8c3f18d570f517acabdaf7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/295418","number":295418,"commit":"e8ba84f042489745aa25a55330e57df9a1212bb7"}],"number":296155}
buildkite-pr-comment-->

flash1293 committed:
fix tests

flash1293 committed:
Merge branch 'flash1293/streams-permissions' of github.com:flash1293/kibana into flash1293/streams-permissions

flash1293 commented:
@kdelemme Do you have Elasticsearch-level permissions to actually create a child stream with your role? The UI is checking both the Kibana and Elasticsearch permissions and only if both are available, it enables features in the UI.

kdelemme commented:
@flash1293 Yes sorry for the false alarm, was just checking the reason and noticed this privileges were coming from the stream response :)
All good.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/296200)
* Commit: 3d8927cb18430294964e4f3ea33256df56364155

### Failed CI Steps
* [Linting](https://buildkite.com/elastic/kibana-pull-request/builds/296200#01966d13-cd45-499a-84d8-743382d676a6)

### History
* :broken_heart: [Build #296155](https://buildkite.com/elastic/kibana-pull-request/builds/296155) failed e5b052a0c49221305bfec9dac31de81e38bccf2a
* :broken_heart: [Build #295890](https://buildkite.com/elastic/kibana-pull-request/builds/295890) failed 726d4976625c3b4ab963859a56a69614e2a19383
* :broken_heart: [Build #295861](https://buildkite.com/elastic/kibana-pull-request/builds/295861) failed 0491ca46ce4e1c54aaaeeb8831cb904e0a786b71
* :broken_heart: [Build #295617](https://buildkite.com/elastic/kibana-pull-request/builds/295617) failed 4d79e68355bb346bea8c3f18d570f517acabdaf7
* :broken_heart: [Build #295418](https://buildkite.com/elastic/kibana-pull-request/builds/295418) failed e8ba84f042489745aa25a55330e57df9a1212bb7

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/296200","number":296200,"commit":"3d8927cb18430294964e4f3ea33256df56364155"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/296155","number":296155,"commit":"e5b052a0c49221305bfec9dac31de81e38bccf2a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/295890","number":295890,"commit":"726d4976625c3b4ab963859a56a69614e2a19383"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/295861","number":295861,"commit":"0491ca46ce4e1c54aaaeeb8831cb904e0a786b71"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/295617","number":295617,"commit":"4d79e68355bb346bea8c3f18d570f517acabdaf7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/295418","number":295418,"commit":"e8ba84f042489745aa25a55330e57df9a1212bb7"}],"number":296200}
buildkite-pr-comment-->

jeramysoucy reviewed (approved):
Thanks for the updates! LGTM

flash1293 committed:
Merge remote-tracking branch 'upstream/main' into flash1293/streams-permissions

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/296380)
* Commit: ec18ef0e41d8ccf7380cd63beadb7936986c3598

### Failed CI Steps
* [x-pack/test/alerting_api_integration/security_and_spaces/group1/config.ts](https://buildkite.com/elastic/kibana-pull-request/builds/296380#01966e51-39bf-4333-b549-6f7c05aba788)
* [FTR Configs #92](https://buildkite.com/elastic/kibana-pull-request/builds/296380#01966e51-3a0c-4432-bbcb-d35b9f17c4cb)
* [FTR Configs #92](https://buildkite.com/elastic/kibana-pull-request/builds/296380#01966e79-8a7b-449b-ad91-003f30600da9)
* [FTR Configs #98](https://buildkite.com/elastic/kibana-pull-request/builds/296380#01966e51-3a12-43ba-8ee9-28e78891d577)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/296380#01966e51-39bf-4333-b549-6f7c05aba788) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/296380/jobs/01966e51-39bf-4333-b549-6f7c05aba788/artifacts/01966e57-3477-47f3-8426-2b86b1c710d7) x-pack/test/alerting_api_integration/security_and_spaces/group1/config.ts / alerting api integration security and spaces enabled Alerts - Group 1 alerts backfill rule runs backfill api key invalidation should wait to invalidate API key until backfill for rule is complete
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/296380#01966e51-3a0c-4432-bbcb-d35b9f17c4cb) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/296380/jobs/01966e51-3a0c-4432-bbcb-d35b9f17c4cb/artifacts/01966e79-63a4-4afe-9ecb-f580824b508e) FTR Configs #<span></span>92 / spaces api with security get superuser should return 200 when referencing the space_3 space explicitly in the URL
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/296380#01966e51-3a12-43ba-8ee9-28e78891d577) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/296380/jobs/01966e51-3a12-43ba-8ee9-28e78891d577/artifacts/01966e78-62f9-4249-8e86-13d90fc91765) FTR Configs #<span></span>98 / spaces api with security get superuser should return 200 when referencing the space_3 space explicitly in the URL
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/296380#01966e79-8a7b-449b-ad91-003f30600da9) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/296380/jobs/01966e79-8a7b-449b-ad91-003f30600da9/artifacts/01966e8a-a243-4d85-b8f2-19cd4c642f79) FTR Configs #<span></span>92 / spaces api with security get superuser should return 200 when referencing the space_3 space explicitly in the URL

### History
* :broken_heart: [Build #296200](https://buildkite.com/elastic/kibana-pull-request/builds/296200) failed 3d8927cb18430294964e4f3ea33256df56364155
* :broken_heart: [Build #296155](https://buildkite.com/elastic/kibana-pull-request/builds/296155) failed e5b052a0c49221305bfec9dac31de81e38bccf2a
* :broken_heart: [Build #295890](https://buildkite.com/elastic/kibana-pull-request/builds/295890) failed 726d4976625c3b4ab963859a56a69614e2a19383
* :broken_heart: [Build #295861](https://buildkite.com/elastic/kibana-pull-request/builds/295861) failed 0491ca46ce4e1c54aaaeeb8831cb904e0a786b71
* :broken_heart: [Build #295617](https://buildkite.com/elastic/kibana-pull-request/builds/295617) failed 4d79e68355bb346bea8c3f18d570f517acabdaf7

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/296380","number":296380,"commit":"ec18ef0e41d8ccf7380cd63beadb7936986c3598"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/296200","number":296200,"commit":"3d8927cb18430294964e4f3ea33256df56364155"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/296155","number":296155,"commit":"e5b052a0c49221305bfec9dac31de81e38bccf2a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/295890","number":295890,"commit":"726d4976625c3b4ab963859a56a69614e2a19383"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/295861","number":295861,"commit":"0491ca46ce4e1c54aaaeeb8831cb904e0a786b71"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/295617","number":295617,"commit":"4d79e68355bb346bea8c3f18d570f517acabdaf7"}],"number":296380}
buildkite-pr-comment-->

flash1293 committed:
Merge remote-tracking branch 'upstream/main' into flash1293/streams-permissions

flash1293 committed:
fix tests

flash1293 committed:
Merge branch 'flash1293/streams-permissions' of github.com:flash1293/kibana into flash1293/streams-permissions

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/296553)
* Commit: f158d0e182c383ddc8e6e79f7d3333d68637dfb4

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/b46e65154ddc7516328cd87b58523809d8e7bfeb) | [after](https://github.com/elastic/kibana/commit/f158d0e182c383ddc8e6e79f7d3333d68637dfb4) | diff |
| --- | --- | --- | --- |
| `streams` | 20 | 21 | +1 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/b46e65154ddc7516328cd87b58523809d8e7bfeb) | [after](https://github.com/elastic/kibana/commit/f158d0e182c383ddc8e6e79f7d3333d68637dfb4) | diff |
| --- | --- | --- | --- |
| `streams` | 11 | 12 | +1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/b46e65154ddc7516328cd87b58523809d8e7bfeb) | [after](https://github.com/elastic/kibana/commit/f158d0e182c383ddc8e6e79f7d3333d68637dfb4) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 405.6KB | 405.9KB | +297.0B |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/b46e65154ddc7516328cd87b58523809d8e7bfeb) | [after](https://github.com/elastic/kibana/commit/f158d0e182c383ddc8e6e79f7d3333d68637dfb4) | diff |
| --- | --- | --- | --- |
| `streams` | 6.6KB | 6.8KB | +223.0B |
| `streamsApp` | 10.1KB | 10.3KB | +136.0B |
| `streamsAppWrapper` | 4.5KB | 4.7KB | +125.0B |
| total |  |  | +484.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/b46e65154ddc7516328cd87b58523809d8e7bfeb) | [after](https://github.com/elastic/kibana/commit/f158d0e182c383ddc8e6e79f7d3333d68637dfb4) | diff |
| --- | --- | --- | --- |
| `streams` | 11 | 12 | +1 |

#### ESLint disabled line counts
| id | [before](https://github.com/elastic/kibana/commit/b46e65154ddc7516328cd87b58523809d8e7bfeb) | [after](https://github.com/elastic/kibana/commit/f158d0e182c383ddc8e6e79f7d3333d68637dfb4) | diff |
| --- | --- | --- | --- |
| `streams` | 2 | 3 | +1 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/b46e65154ddc7516328cd87b58523809d8e7bfeb) | [after](https://github.com/elastic/kibana/commit/f158d0e182c383ddc8e6e79f7d3333d68637dfb4) | diff |
| --- | --- | --- | --- |
| `streams` | 5 | 6 | +1 |


</details>


### History
* :broken_heart: [Build #296380](https://buildkite.com/elastic/kibana-pull-request/builds/296380) failed ec18ef0e41d8ccf7380cd63beadb7936986c3598
* :broken_heart: [Build #296200](https://buildkite.com/elastic/kibana-pull-request/builds/296200) failed 3d8927cb18430294964e4f3ea33256df56364155
* :broken_heart: [Build #296155](https://buildkite.com/elastic/kibana-pull-request/builds/296155) failed e5b052a0c49221305bfec9dac31de81e38bccf2a
* :broken_heart: [Build #295890](https://buildkite.com/elastic/kibana-pull-request/builds/295890) failed 726d4976625c3b4ab963859a56a69614e2a19383
* :broken_heart: [Build #295861](https://buildkite.com/elastic/kibana-pull-request/builds/295861) failed 0491ca46ce4e1c54aaaeeb8831cb904e0a786b71

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/296553","number":296553,"commit":"f158d0e182c383ddc8e6e79f7d3333d68637dfb4"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/296380","number":296380,"commit":"ec18ef0e41d8ccf7380cd63beadb7936986c3598"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/296200","number":296200,"commit":"3d8927cb18430294964e4f3ea33256df56364155"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/296155","number":296155,"commit":"e5b052a0c49221305bfec9dac31de81e38bccf2a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/295890","number":295890,"commit":"726d4976625c3b4ab963859a56a69614e2a19383"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/295861","number":295861,"commit":"0491ca46ce4e1c54aaaeeb8831cb904e0a786b71"}],"number":296553}
buildkite-pr-comment-->

kibanamachine commented:
Starting backport for target branches: 8.19

https://github.com/elastic/kibana/actions/runs/14704943899

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.19|Backport failed because of merge conflicts|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 218966
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->
